lib/functions/installs.bash

```diff
@@ -117,30 +121,146 @@ install_local_tool_versions() {
     exit 1
   fi
 
+  local tools_installed
   if [ -n "$plugins_installed" ]; then
-    for plugin_name in $plugins_installed; do
-      local plugin_version_and_path
-      plugin_version_and_path="$(find_versions "$plugin_name" "$search_path")"
-
-      if [ -n "$plugin_version_and_path" ]; then
-        local plugin_version
-        some_tools_installed='yes'
-        plugin_versions=$(cut -d '|' -f 1 <<<"$plugin_version_and_path")
-        for plugin_version in $plugin_versions; do
-          install_tool_version "$plugin_name" "$plugin_version"
-        done
-      fi
-    done
+    plugins_installed=$(printf "%s" "$plugins_installed" | tr "\n" " " | awk '{$1=$1};1')
+    display_debug "install_local_tool_versions: plugins_installed='$plugins_installed'"
+    tools_installed=$(install_directory_tools_recursive "$search_path" "$plugins_installed")
   fi
 
-  if [ -z "$some_tools_installed" ]; then
+  if [ -z "$tools_installed" ]; then
     printf "Either specify a tool & version in the command\n"
     printf "OR add .tool-versions file in this directory\n"
     printf "or in a parent directory\n"
     exit 1
   fi
 }
 
+install_directory_tools_recursive() {
+  local search_path=$1
+  local plugins_installed=$2
+  local tools_installed=""
+
+  display_debug "install_directory_tools_recursive '$search_path': entered with plugins_installed='$plugins_installed'"
+
+  while [ "$search_path" != "/" ]; do
+    # install tools from files in current directory
+    display_debug "--------------------------------------------------------------------------------------------------------------"
+    tools_installed=$(install_directory_tools "$search_path" "$plugins_installed" "$tools_installed")
+    display_debug "install_directory_tools_recursive '$search_path': install_directory_tools returned tools_installed='$tools_installed'"
+
+    # terminate if all tools are installed
+    if [[ -n $(stringlist_same_length "$plugins_installed" "$tools_installed") ]]; then
```

Comment:
perhaps the validation should be tools_installed contains all entries in plugins_installed... this would be a bit stronger